### PR TITLE
[WIP] Profiling of phases

### PR DESCRIPTION
**Pull request by InTheCloudDan**
My first attempt at: https://github.com/facebook/prepack/issues/413

I'm not sure if all of the console.time's are in the correct spots.

From the issue, I don't know where to find: How long does speculative initialization of more modules take?

### CODE REVIEW

**Comment by InTheCloudDan** [comment]
I'm not sure how to interpret the test. Seems like a whole lot of timeouts, is there a way to force it to retry?

File: src/serializer/serializer.js [hunk 1]
```diff
@@ -1777,26 +1776,32 @@ export class Serializer {
...
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Interpret");
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
     if (this.options.initializeMoreModules) {
       this.modules.initializeMoreModules();
       if (this.logger.hasErrors()) return undefined;
     }
+    if (this.options.profile) console.timeEnd("Interpret");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This should move to just before the if above.

File: src/serializer/serializer.js [hunk 2]
```diff
@@ -1777,26 +1776,32 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Interpret");
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
     if (this.options.initializeMoreModules) {
       this.modules.initializeMoreModules();
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is where the speculative initialization happens.

**Comment by hermanventer** [comment]
The CircleCI page has a rebuild button. I've triggered another build. The time-outs might be due to load on the CircleCI servers.

File: src/serializer/serializer.js [hunk 3]
```diff
@@ -1777,9 +1776,11 @@ export class Serializer {
   init(filename: string, code: string, map?: string = "",
       sourceMaps?: boolean = false, onError?: (Realm, Value) => void) {
     // Phase 1: Let's interpret.
+    if (this.options.profile) console.time("Speculative Initialization");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is the interpretation of the first initialization phase.

File: src/serializer/serializer.js [hunk 4]
```diff
@@ -1777,9 +1776,11 @@ export class Serializer {
...
     this.execute(filename, code, map, onError);
     if (this.logger.hasErrors()) return undefined;
     this.modules.resolveInitializedModules();
+    if (this.options.profile) console.timeEnd("Speculative Initialization");
     if (this.options.initializeMoreModules) {
       this.modules.initializeMoreModules();
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is speculative initialization of more modules.

File: src/serializer/serializer.js [hunk 5]
```diff
@@ -1788,15 +1789,19 @@ export class Serializer {
     // Phase 2: Let's serialize the heap and generate code.
     // Serialize for the first time in order to gather reference counts
     if (!this.options.singlePass) {
+      if (this.options.profile) console.time("Second Serialize Pass");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is the first (but optional) pass.

File: src/serializer/serializer.js [hunk 6]
```diff
@@ -1788,15 +1789,19 @@ export class Serializer {
...
       this.collectValToRefCountOnly = true;
       this.valToRefCount = new Map();
       this.serialize(filename, code, sourceMaps);
       if (this.logger.hasErrors()) return undefined;
+      if (this.options.profile) console.timeEnd("Second Serialize Pass");
     }
     // Serialize for a second time, using reference counts to minimize number of generated identifiers
     this._resetSerializeStates();
     this.collectValToRefCountOnly = false;
+    if (this.options.profile) console.time("First Serialize Pass");
```
Comment (hermanventer) [inline src/serializer/serializer.js]: This is the second serialization pass.

**Comment by Kishore-B-Rao** [comment]
See Profiling #714 for continuation of this pull request

**Pull request merged**